File: sdk-ts/src/functions/remote.ts
import { Client } from "@hey-api/client-fetch";
import { StructuredTool, tool } from "@langchain/core/tools";
import { Client as ModelContextProtocolClient } from "@modelcontextprotocol/sdk/client/index.js";
import { Transport } from "@modelcontextprotocol/sdk/shared/transport.js";
import { z } from "zod";
import { getAuthenticationHeaders } from "../authentication/authentication.js";
import { getFunction, listFunctions } from "../client/sdk.gen.js";
import { Function } from "../client/types.gen.js";
import { getSettings, Settings } from "../common/settings.js";
import { RunClient } from "../run.js";
import { parametersToZodSchema } from "./common.js";
import { MCPClient, MCPToolkit } from "./mcp.js";
import { WebSocketClientTransport } from "./transport/websocket.js";

/**
 * Creates a StructuredTool for remote functions, enabling their invocation via the RunClient.
 *
 * @param {RunClient} client - The client instance used to execute the function.
 * @param {string} name - The name of the remote function.
 * @param {string} description - A description of what the function does.
 * @param {z.ZodType} schema - The Zod schema for the function's input parameters.
 * @returns {StructuredTool} The structured tool representing the remote function.
 */
export function getRemoteTool(
  client: RunClient,
  name: string,
  description: string,
  schema: z.ZodType
) {
  return tool(
    async (args: Record<string, any>) => {
      const data = await client.run({
        resourceType: "function",
        resourceName: name,
        method: "POST",
        json: args,
      });
      return data;
    },
    {
      name,
      description,
      schema,
    }
  );
}

/**
 * Toolkit for managing and interacting with remote toolkits and MCP services.
 */
export class RemoteToolkit {
  private client: Client;
  private modelContextProtocolClient: ModelContextProtocolClient;
  private _mcpToolkit: MCPToolkit | null = null;
  private fallbackUrl: string | null = null;
  private functionName: string;
  private _function: Function | null = null;
  private runClient: RunClient;
  private settings: Settings;

  /**
   * Creates an instance of RemoteToolkit.
   *
   * @param {Client} client - The HTTP client instance.
   * @param {string} functionName - The name of the remote function to manage.
   */
  constructor(client: Client, functionName: string) {
    this.settings = getSettings();
    this.client = client;
    this.functionName = functionName;
    this.runClient = new RunClient(client);
    this.modelContextProtocolClient = new ModelContextProtocolClient(
      {
        name: this.settings.name,
        version: "1.0.0",
      },
      {
        capabilities: {
          tools: {},
        },
      }
    );
  }

  /**
   * Initializes the toolkit by retrieving the specified function and its associated tools.
   *
   * @returns {Promise<void>} Resolves when initialization is complete.
   * @throws Will throw an error if the function retrieval fails.
   */
  async initialize(): Promise<void> {
    if (!this._function) {
      const { response, data } = await getFunction({
        client: this.client,
        path: { functionName: this.functionName },
      });
      if (response.status >= 400) {
        const { data: listData } = await listFunctions({
          client: this.client,
        });
        const names =
          listData?.map((f: Function) => f.metadata?.name || "") || [];
        throw new Error(
          `error ${response.status}. Available functions: ${names.join(", ")}`
        );
      }
      this._function = data || null;
    }

    if (
      this._function &&
      this._function.metadata &&
      this._function.spec?.integrationConnections
    ) {
      let url = `${this.settings.runUrl}/${this.settings.workspace}/functions/${this._function.metadata.name}`;
      let transport: Transport;
      const headers = await getAuthenticationHeaders();
      const envVar = toEnvVar(this._function.metadata.name || "");
      if (process.env[`BL_FUNCTION_${envVar}_SERVICE_NAME`]) {
        this.fallbackUrl = url;
        url = `https://${process.env[`BL_FUNCTION_${envVar}_SERVICE_NAME`]}.${
          this.settings.runInternalHostname
        }`;
        transport = new WebSocketClientTransport(new URL(url), {
          "x-beamlit-authorization": headers?.["X-Beamlit-Authorization"] || "",
          "x-beamlit-workspace": headers?.["X-Beamlit-Workspace"] || "",
        });
      } else {
        transport = new WebSocketClientTransport(new URL(url), {
          "x-beamlit-authorization": headers?.["X-Beamlit-Authorization"] || "",
          "x-beamlit-workspace": headers?.["X-Beamlit-Workspace"] || "",
        });
      }
      try {
        await this.modelContextProtocolClient.connect(transport);
        const mcpClient = new MCPClient(this.modelContextProtocolClient, transport);
        const mcpToolkit = new MCPToolkit(mcpClient);
        this._mcpToolkit = mcpToolkit;
        await mcpToolkit.initialize();
      } catch (error) {
        if (this.fallbackUrl) {
          transport = new WebSocketClientTransport(new URL(this.fallbackUrl), {
            "x-beamlit-authorization":
              headers?.["X-Beamlit-Authorization"] || "",
            "x-beamlit-workspace": headers?.["X-Beamlit-Workspace"] || "",
          });
          await this.modelContextProtocolClient.connect(transport);
          const mcpClient = new MCPClient(this.modelContextProtocolClient, transport);
          const mcpToolkit = new MCPToolkit(mcpClient);
          this._mcpToolkit = mcpToolkit;
          try {
            await mcpToolkit.initialize();
          } catch (error) {
            throw new Error(
              `Failed to initialize MCP toolkit, error: ${error}`
            );
          }
        } else {
          throw new Error(`Failed to initialize MCP toolkit, error: ${error}`);
        }
      }
    }
  }

  /**
   * Retrieves the list of structured tools from the remote function. If the function has integration connections,
   * it utilizes the MCPToolkit to manage them.
   *
   * @returns {Promise<StructuredTool[]>} An array of structured tools.
   * @throws Will throw an error if the toolkit has not been initialized.
   */
  async getTools(): Promise<StructuredTool[]> {
    if (!this._function) {
      throw new Error("Must initialize the toolkit first");
    }
    if (this._mcpToolkit) {
      return this._mcpToolkit.getTools();
    }
    return [
      getRemoteTool(
        this.runClient,
        this._function.metadata?.name || "",
        this._function.spec?.description || "",
        parametersToZodSchema(this._function.spec?.parameters || [])
      ),
    ];
  }
}

function toEnvVar(name: string) {
  return name.replace(/-/g, "_").toUpperCase();
}

File: sdk-python/src/blaxel/functions/remote/remote.py
"""
This module provides functionalities to integrate remote functions into Blaxel.
It includes classes for creating dynamic schemas based on function parameters and managing remote toolkits.
"""

import asyncio
import os
import warnings
from dataclasses import dataclass
from typing import Callable

import pydantic
import typing_extensions as t
from langchain_core.tools.base import BaseTool, ToolException

from blaxel.api.functions import get_function, list_functions
from blaxel.authentication.authentication import AuthenticatedClient
from blaxel.common.settings import get_settings
from blaxel.errors import UnexpectedStatus
from blaxel.functions.mcp.mcp import MCPClient, MCPToolkit
from blaxel.models import Function, StoreFunctionParameter
from blaxel.run import RunClient


def create_dynamic_schema(name: str, parameters: list[StoreFunctionParameter]) -> type[pydantic.BaseModel]:
    """
    Creates a dynamic Pydantic schema based on function parameters.

    Args:
        name (str): The name of the schema.
        parameters (list[StoreFunctionParameter]): List of parameter objects.

    Returns:
        type[pydantic.BaseModel]: The dynamically created Pydantic model.
    """
    field_definitions = {}
    for param in parameters:
        field_type = str
        if param.type_ == "number":
            field_type = float
        elif param.type_ == "integer":
            field_type = int
        elif param.type_ == "boolean":
            field_type = bool

        field_definitions[param.name] = (
            field_type,
            pydantic.Field(description=param.description or "")
        )
    return pydantic.create_model(
        f"{name}Schema",
        **field_definitions
    )


class RemoteTool(BaseTool):
    """
    Tool for interacting with remote functions.

    Attributes:
        client (RunClient): The client used to execute remote function calls.
        resource_name (str): The name of the remote resource.
        kit (bool): Indicates whether the tool is part of a function kit.
        handle_tool_error (bool | str | Callable[[ToolException], str] | None): Error handling strategy.
    """

    client: RunClient
    resource_name: str
    kit: bool = False
    handle_tool_error: bool | str | Callable[[ToolException], str] | None = True
    service_name: str | None = None
    cloud: bool = False
    @t.override
    def _run(self, *args: t.Any, **kwargs: t.Any) -> t.Any:
        warnings.warn(
            "Invoke this tool asynchronousely using `ainvoke`. This method exists only to satisfy standard tests.",
            stacklevel=1,
        )
        return asyncio.run(self._arun(*args, **kwargs))

    @t.override
    async def _arun(self, *args: t.Any, **kwargs: t.Any) -> t.Any:
        body = {**kwargs}
        if self.kit:
            body["name"] = self.name
        result = self.client.run(
            "function",
            self.resource_name,
            "POST",
            cloud=self.cloud,
            service_name=self.service_name,
            json=body,
        )
        return result.text

    @t.override
    @property
    def tool_call_schema(self) -> type[pydantic.BaseModel]:
        assert self.args_schema is not None  # noqa: S101
        return self.args_schema

@dataclass
class RemoteToolkit:
    """
    Toolkit for managing remote function tools.

    Attributes:
        client (AuthenticatedClient): The authenticated client instance.
        function (str): The name of the remote function to integrate.
        _function (Function | None): Cached Function object after initialization.
    """
    client: AuthenticatedClient
    function: str
    _function: Function | None = None
    _service_name: str | None = None
    model_config = pydantic.ConfigDict(arbitrary_types_allowed=True)

    async def initialize(self) -> None:
        """Initialize the session and retrieve the remote function details."""
        if self._function is None:
            try:
                response = get_function.sync_detailed(self.function, client=self.client)
                function_name = self.function.upper().replace("-", "_")
                if os.getenv(f"BL_FUNCTION_{function_name}_SERVICE_NAME"):
                    self._service_name = os.getenv(f"BL_FUNCTION_{function_name}_SERVICE_NAME")
                self._function = response.parsed
            except UnexpectedStatus as e:
                functions = list_functions.sync_detailed(
                    client=self.client,
                ).parsed
                names = [
                    f.metadata.name
                    for f in functions
                ]
                raise RuntimeError(
                    f"error: {e.status_code}. Available functions: {', '.join(names)}"
                )

    async def get_tools(self) -> list[BaseTool]:
        settings = get_settings()
        if self._function is None:
            raise RuntimeError("Must initialize the toolkit first")

        if self._function.spec.integration_connections:
            fallback_url = None
            url = f"{settings.run_url}/{settings.workspace}/functions/{self._function.metadata.name}"
            if self._service_name:
                fallback_url = f"https://{self._service_name}.{settings.run_internal_hostname}"
                url = f"https://{self._service_name}.{settings.run_internal_hostname}"
            mcp_client = MCPClient(self.client, url, fallback_url)
            mcp_toolkit = MCPToolkit(client=mcp_client, url=url)
            await mcp_toolkit.initialize()
            return await mcp_toolkit.get_tools()

        if self._function.spec.kit:
            return [
                RemoteTool(
                    client=RunClient(self.client),
                    name=func.name,
                    resource_name=self._function.metadata.name,
                    kit=True,
                    description=func.description or "",
                    args_schema=create_dynamic_schema(func.name, func.parameters),
                    cloud=settings.cloud,
                    service_name=self._service_name,
                )
                for func in self._function.spec.kit
            ]

        return [
            RemoteTool(
                client=RunClient(self.client),
                name=self._function.metadata.name,
                resource_name=self._function.metadata.name,
                description=self._function.spec.description or "",
                args_schema=create_dynamic_schema(
                    self._function.metadata.name,
                    self._function.spec.parameters
                ),
                cloud=settings.cloud,
                service_name=self._service_name,
            )
        ]